### ISSUE TITLE
cleanup start.d usage in distro

### ISSUE THREAD

**Issue by gregw** [eclipse/jetty.project#589]
stop using a mixed style of start.ini and start.d in the distro

**Comment by joakime** [comment]
How about removing start.ini from ${jetty.home} too.
Make Jetty 10+ not capable of being run from ${jetty.home} directly.

**Comment by gregw** [comment]
Also we could make add-to-start check for the existence of a start.d and if it exists users it.  Similarly add-to-startd would warn or fail if there is a start.ini

We could then just document examples using add to start.

**Comment by jmcc0nn3ll** [comment]
Like @WalkerWatch was saying, ultimately this should be the goal and the documentation and stock distribution should all clearly adhere to one approach, and then have the alternative documented in its own section.

I just wish I wasn't on the fence as to which is best to lead with.  I love the ability to pop open one file and control everything, but it has grown to a large file which makes it unwieldy.

**Comment by jmcc0nn3ll** [comment]
@gregw Perhaps the proper approach here is to offer two distributions.  This goes well with something @joakime and I were talking about the other day.  As we push more things into the jetty-distribution artifact, like the documentation I am going to be wiring in shortly, we ought to provide a trimmed down offering that is more suitable as a building block or *cough* power user.  Something appropriate for the various docker containers, and that use case the start.d is highly suitable.

So I am kinda thinking that we ought to orient the jetty-distribution artifact for the end user and commit to the start.ini approach.  It is the simplest case to explain... And then create a jetty-distribution-light or obviously something with a better name that is more suited for people building out their own solution where layering config makes tons more sense.

**Comment by WalkerWatch** [comment]
@gregw after some thought and then a discussion with @jmcc0nn3ll I could see two distros as a way forward - one for a more typical Jetty roll out, utilizing a single start.ini and another for more complex configuration using the start.d directory. The majority of users will likely use the former, and documentation can be focused that way with additional sections on how to utilize start.d for managing environments with more moving pieces.

**Comment by jmcc0nn3ll** [comment]
I pushed a branch (distro-rework-589) with the changes we discussed for jetty-9.4.x.

I created a jetty-home artifact that by itself works as a jetty.home, you can run 'java -jar path/to/start.jar --add-to-startd=http' and it will populate a base that can run.  In of itself it can't be used as a distribution, no webapps dir, no start.ini, just all the elements of a proper jetty.home to build jetty.bases from.

I updated the jetty-distribution so that it first pulls in the jetty-home and then populates it with the rest of the trappings of the traditional jetty-distribution.  The net affect is both modules have smaller poms and have different purposes...seems to work quite nicely.

Additionally, the jetty-distribution now only uses the --add-to-start=<module> exec's to flesh out the distribution.  I am ultimately fine with one or the other...but I do not thing our jetty-distribution or its demo-base should mix these concepts...and the jetty-documentation should be written against the jetty-distribution for examples and walkthroughs.

thoughts?

**Comment by gregw** [comment]
@jmcc0nn3ll you should create a PR just so we can use the tools to easily see the diffs and comment line by line.

But in this case, I think this particular change is widely accepted so if it works, you may as well merge to the 9.4.x branch (it is not like that currently builds anyway:(

As for the the start.ini vs start.d choice, I agree that we should never mix styles.       I'm thinking that as the jetty-distribution might be used as the basis for other projects, then it should use start.d as that makes is simpler to remove/replace the configuration of an individual module.  Demo-base can be setup to use a single start.ini

**Comment by jmcc0nn3ll** [comment]
@gregw Well my intention was that if another project was to build off of a distribution then they would use jetty-home just like we do in jetty-distribution and then run the relevant --add-to-startd= commands to build out it out however they like.  Currently jetty-distribution (as a jetty.home and jetty.base) is completely  start.ini and the demo-base is the hybrid mixing start.ini and start.d configuration.  With the intention of making the documentation consistent in how it discusses configuration, I would think in the jetty-distribution both demo-base and jetty-distribution should be just start.ini based.

Alternately, we could tease out one or two of the demo webapps and make a demo-base-alt that we configure using start.d and then in some documentation discussing those we can introduce better the start.d alternative configuration approach.  Or perhaps we make a doc-base that publishes the documentation using start.d?

Primary goal here being for the distribution+documentation combo we present a clean consistent story to the entry level user.  Secondary goal (jetty-home artifact) being that we provide a useable artifact to build custom jetty distributions from.

**Comment by gregw** [comment]
@jmcc0nn3ll I think we agreed that all the doco should be just -add-to-start    and that we'd implement the feature that if a start.d existed then it would be used.   So let's not complicate and/or fork the doco and stick with that.

Making the documentation consistent does not need to have either start.ini or start.d, as it just uses --add-to-start and how that is stored is irrelevant.    Hmm I guess we have to talk about editing the generated file sometimes, but then that argues for using start.d because it is so much easier to say "edit start.d/foobar.ini" then it is to say "edit start.ini and find the foobar section in it and edit that".

So the doco should use --add-to-start   everywhere and we need a standard form of words/instructions for editing the configuration associated with a module that is independent of which mechanism is used.  Maybe we need another option on start.jar like  --show-ini=foobar that tells which file is the ini for a module.

I agree that because jetty-home now exists, it's not so important that we make jetty-distribution extensible....    but then it might eventually be extended as another distribution.     

I think that we should demonstrate both styles in the distribution, so demo-base should use one style and the jetty-home should use the other.   Because the distribution might be extended, I think using start.d in the home is best and use a start.ini in the demo-base.

Also we should modify start.jar so that if it detects it is running in jetty-home but there are no ini files of either variety that it gives a very good error message saying that a jetty-base should be created with --add-to-start

In summary:

- No to forking the doco or making any Alt versions of anything
- Use --add-to-start in all the doco
- Find a common form of words (perhaps with extra start.jar option) to describe which file to edit when configuring a module that works for either mechanism
- No to mixing styles
- Use start.d for jetty-distribution home
- Use start.ini for demo-base

**Comment by jmcc0nn3ll** [comment]
I didn't mean a forking of the documentation, just that we could have two demo-bases in the jetty-distribution, the primary one that we reference in the documentation, and then another with one or two things in it that illustrate the start.d configuration approach.

Will process the rest later, just wanted to make sure it was clear I wasn't suggesting forking docs...ewe!

**Comment by gregw** [comment]
I'm not so keen on multiple demo-bases either... at least not only 2.  If we wanted multiples, then we should have lots of different configurations to show.   But then the doco should eventually have lots of simple examples to follow to create demo bases.

So one-demo base is best I think and it can use start.ini (it has custom stuff in there anyway).     Because there is quiet a lot in there, I think we should discourage people looking too closely at demo-base and instead have more simple examples in the doco of how to build bases with specific features.

Demo-base is really more of a welcome-to-jetty-base rather than an example of how to configure a base....    If we want it to be a good example, we should probably create a demo module that adds the extra stuff needed rather than crafting a start.ini

**Comment by jmcc0nn3ll** [comment]
well, if we _have_ to have the distribution contain both configuration styles then I think we should move the hybrid demo-base to be strictly start.d and leave the jetty-distribution itself as the start.ini it has been for the last several years.  I know it makes it harder for people layering with the jetty distribution in the long run but it is no different then it is today, and we are adding the jetty-home option that they can build from alternatively now.

**Comment by jmcc0nn3ll** [comment]
i have updated my branch with your changes but there is a start.ini creeping in from somewhere inside of the demo-base, sorting that out and will merge it over

**Comment by jmcc0nn3ll** [comment]
merged it into 9.4.x

**Comment by gregw** [comment]
See bug report in #644

**Comment by gregw** [comment]
@jmcc0nn3ll I'm trying to reproduce the problem in demo-base and I can't get --add-to-start to create a start.ini ?

Can you try again and step me through what you did to reproduce?

**Comment by jmcc0nn3ll** [comment]
and now I am not seeing it, must have been some old build artifacts or something, if I can get it to happen again I'll reopen
